File: safety/extensions/close-keywords.ts
import type { ExtensionAPI, ExtensionToolCallEvent } from "@oh-my-pi/pi-coding-agent";

/**
 * Rewrite `gh pr create|edit --body` so GitHub close-keywords apply to every
 * issue in a comma list. Advisory-only in Claude; here we rewrite the command
 * input (fail-open) because OMP has no PreToolUse additionalContext channel.
 *
 * The pre-commit commit-msg rewriter is omitted: OMP has no git commit-msg hook.
 */

const KEYWORDS = new Set([
	"close",
	"closes",
	"closed",
	"fix",
	"fixes",
	"fixed",
	"resolve",
	"resolves",
	"resolved",
]);

const REF = /^(?:[A-Za-z0-9_.-]+\/[A-Za-z0-9_.-]+)?#[0-9]+|GH-[0-9]+/i;
const WORD = /^[A-Za-z]+/;
const SPACE = /^[^\S\n]+/;
const SEPARATORS = [
	/^[^\S\n]*,[^\S\n]+and[^\S\n]+/,
	/^[^\S\n]+and[^\S\n]+/,
	/^[^\S\n]*,[^\S\n]*/,
];

const MAX_COMMAND_LENGTH = 64_000;

function matchAt(pattern: RegExp, text: string, pos: string | number): string | null {
	const slice = typeof pos === "number" ? text.slice(pos) : pos;
	const m = pattern.exec(slice);
	return m && m.index === 0 ? m[0] : null;
}

function separatorAt(text: string, pos: number): string | null {
	const slice = text.slice(pos);
	for (const pattern of SEPARATORS) {
		const found = matchAt(pattern, slice, 0);
		if (found) return found;
	}
	return null;
}

export function normalizeLine(line: string): string {
	const out: string[] = [];
	let pos = 0;
	const end = line.length;
	while (pos < end) {
		const last = out.length ? out[out.length - 1] : "";
		const atBoundary = !last || !(/[A-Za-z0-9_]$/.test(last));
		const word = atBoundary ? matchAt(WORD, line, pos) : null;
		if (word === null) {
			out.push(line[pos]);
			pos += 1;
			continue;
		}
		if (!KEYWORDS.has(word.toLowerCase())) {
			out.push(word);
			pos += word.length;
			continue;
		}
		const afterWord = pos + word.length;
		const space = matchAt(SPACE, line, afterWord) ?? "";
		const firstRef = matchAt(REF, line, afterWord + space.length);
		if (firstRef === null) {
			out.push(word);
			pos += word.length;
			continue;
		}
		const keyword = word.toLowerCase();
		out.push(word + space + firstRef);
		pos = afterWord + space.length + firstRef.length;
		while (true) {
			const separator = separatorAt(line, pos);
			if (separator === null) break;
			const ref = matchAt(REF, line, pos + separator.length);
			if (ref === null) break;
			out.push(`${separator}${keyword} ${ref}`);
			pos += separator.length + ref.length;
		}
	}
	return out.join("");
}

export function normalize(text: string): string {
	return text.split("\n").map(normalizeLine).join("\n");
}

type Token = { value: string; start: number; end: number };

// Only literal shell words are eligible. Unsupported shell syntax is never rewritten.
function literalSegments(command: string): Token[][] | null {
	const segments: Token[][] = [[]];
	let i = 0;
	while (i < command.length) {
		if (" \t\r".includes(command[i])) { i++; continue; }
		if (";&|\n".includes(command[i])) {
			const ch = command[i++];
			if ((ch === "&" || ch === "|") && command[i] === ch) i++;
			segments.push([]);
			continue;
		}
		const start = i;
		let value = "";
		let quote: string | null = null;
		while (i < command.length) {
			const ch = command[i];
			if (!quote && " \t\r;&|\n".includes(ch)) break;
			if (ch === "'" && quote !== '"') { quote = quote ? null : "'"; i++; continue; }
			if (ch === '"' && quote !== "'") { quote = quote ? null : '"'; i++; continue; }
			if (quote !== "'" && ch === "\\") {
				const next = command[i + 1];
				if (next === undefined) return null;
				if (quote === '"' && !'$`"\\\n'.includes(next)) {
					value += ch; i++; continue;
				}
				if (next !== "\n") value += next;
				i += 2;
				continue;
			}
			if (quote !== "'" && ("$`".includes(ch) || (!quote && ("()<>*?{}[]~".includes(ch) || (ch === "#" && i === start))))) return null;
			value += ch;
			i++;
		}
		if (quote) return null;
		segments[segments.length - 1].push({ value, start, end: i });
	}
	return segments;
}


function bodySpan(command: string): Token | null {
	const segments = literalSegments(command);
	if (!segments) return null;
	let selected: Token | null = null;
	for (const segment of segments) {
		let start = 0;
		while (/^[A-Za-z_][A-Za-z0-9_]*=/.test(segment[start]?.value ?? "")) start++;
		if (segment[start]?.value !== "gh" || segment[start + 1]?.value !== "pr" ||
			!["create", "edit"].includes(segment[start + 2]?.value ?? "")) continue;
		selected = null;
		for (let i = start + 3; i < segment.length; i++) {
			const token = segment[i];
			if (token.value === "--") break;
			if (token.value === "--body" || token.value === "-b") {
				selected = segment[++i] ?? null;
			} else {
				if (["--title", "-t", "--base", "-B", "--head", "-H", "--repo", "-R",
					"--reviewer", "-r", "--assignee", "-a", "--label", "-l", "--project", "-p",
					"--milestone", "-m", "--body-file", "-F", "--template", "-T",
					"--add-assignee", "--remove-assignee", "--add-label", "--remove-label",
					"--add-project", "--remove-project", "--add-reviewer", "--remove-reviewer"].includes(token.value)) {
					i++;
					continue;
				}
				const prefix = token.value.startsWith("--body=") ? "--body=" :
					token.value.startsWith("-b=") ? "-b=" :
					token.value.startsWith("-b") && !token.value.startsWith("--") ? "-b" : null;
				if (prefix) {
					// Replace the entire word, keeping the flag literal even when originally quoted.
					selected = { ...token, value: token.value.slice(prefix.length) };
				}
			}
		}
	}
	return selected;
}

export function extractBody(command: string): string {
	return bodySpan(command)?.value ?? "";
}

export function replaceLastBody(command: string, next: string): string | null {
	const span = bodySpan(command);
	if (!span) return null;
	const raw = command.slice(span.start, span.end);
	const token = literalSegments(raw)?.[0]?.[0]?.value ?? "";
	const prefix = token !== span.value
		? token.slice(0, token.length - span.value.length)
		: "";
	const escaped = `'${next.replaceAll("'", "'\\''")}'`;
	return command.slice(0, span.start) + prefix + escaped + command.slice(span.end);
}

function commandOf(event: ExtensionToolCallEvent): string {
	const raw = event.input;
	if (typeof raw.command === "string") return raw.command;
	return "";
}

export default function closeKeywords(pi: ExtensionAPI): void {
	pi.on("tool_call", (event) => {
		try {
			if (event.toolName !== "bash") return;
			const command = commandOf(event);
			if (!command || command.length > MAX_COMMAND_LENGTH) return;
			if (!command.includes("gh") || !command.includes("pr")) return;
			const body = extractBody(command);
			if (!body) return;
			const fixed = normalize(body);
			if (fixed === body) return;
			const rewritten = replaceLastBody(command, fixed);
			if (!rewritten) return;
			return { input: { ...event.input, command: rewritten } };
		} catch {
			return;
		}
	});
}

File: project-setup/skills/project-setup/assets/hooks/scripts/close_keywords.py
#!/usr/bin/env python3
# VENDORED from srobroek/agentic-packages, packages/hooks-close-keywords/scripts/close_keywords.py.
#
# Vendored rather than referenced because a prek `entry:` has to resolve inside this
# repository. An installed package does not: it may live anywhere on the machine, or
# nowhere. The package's own templates/pre-commit-commit-msg.yaml names project-setup as
# the consumer that does this.
#
# NO CHECKER ENFORCES THIS COPY. Both repositories have twin-script checks and neither
# reaches across a repository boundary, so drift here is silent. Re-sync with:
#
#     cp <agentic-packages>/packages/hooks-close-keywords/scripts/close_keywords.py \
#        templates/quality/hooks/template/scripts/close_keywords.py
#
# Fix behaviour upstream and re-copy; editing here is reverted by the next sync.
"""Distribute a GitHub closing keyword across a contiguous list of issue refs.

GitHub binds a closing keyword to only the FIRST reference in a list, so
`Closes #37, #38, #39` closes #37 and leaves the rest open. This rewrites such a
list to `Closes #37, closes #38, closes #39`, which closes all three.

Shared engine for both delivery layers: the pre-commit `commit-msg` hook that
rewrites a message in place, and the `PreToolUse` guard that advises on a
`gh pr create` body.

Scope is deliberately narrow, because a false rewrite of someone's prose costs
more than a missed distribution:

- Only a list IMMEDIATELY following a keyword is distributed. An unrelated `#N`
  later on the line is untouched.
- Separators inside the list are `,`, `, `, ` and `, and `, and `.
- Reference forms are `#N`, `owner/repo#N`, and `GH-N`.
- Distributed copies use the lowercased keyword; the original keeps its case.

Idempotent: a list whose refs each already carry a keyword is left alone,
because a separator followed by a keyword rather than a reference ends the list.
"""

from __future__ import annotations

import re

# GitHub's closing keyword set. Matched case-insensitively, on a word boundary,
# so `closet #1, #2` is prose rather than a close directive.
KEYWORDS = frozenset(
    {
        "close",
        "closes",
        "closed",
        "fix",
        "fixes",
        "fixed",
        "resolve",
        "resolves",
        "resolved",
    }
)

_WORD = re.compile(r"[A-Za-z]+")

# `owner/repo#123` or `#123`, else `GH-123`. Anchored: callers test the position
# directly after a keyword or separator, never search forward from it.
_REF = re.compile(r"(?:[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+)?#[0-9]+|GH-[0-9]+", re.IGNORECASE)

# List separators, longest form first so `, and ` is not read as a bare `,`.
_SEPARATORS = (
    re.compile(r"[^\S\n]*,[^\S\n]+and[^\S\n]+"),
    re.compile(r"[^\S\n]+and[^\S\n]+"),
    re.compile(r"[^\S\n]*,[^\S\n]*"),
)


def _match_at(pattern: re.Pattern[str], text: str, pos: int) -> str | None:
    """Return the text `pattern` matches starting exactly at `pos`, if any."""
    match = pattern.match(text, pos)
    return match.group(0) if match else None


def _separator_at(text: str, pos: int) -> str | None:
    for pattern in _SEPARATORS:
        found = _match_at(pattern, text, pos)
        if found:
            return found
    return None


def normalize_line(line: str) -> str:
    """Distribute close keywords across every contiguous list in one line."""
    out: list[str] = []
    pos = 0
    end = len(line)

    while pos < end:
        # A keyword only counts on a word boundary, which is what stops the
        # `closes` inside `precloses` or `closet` from opening a list.
        at_boundary = not out or not (out[-1][-1:].isalnum() or out[-1].endswith("_"))
        word = _match_at(_WORD, line, pos) if at_boundary else None

        if word is None:
            out.append(line[pos])
            pos += 1
            continue

        if word.lower() not in KEYWORDS:
            out.append(word)
            pos += len(word)
            continue

        after_word = pos + len(word)
        space = _match_at(re.compile(r"[^\S\n]+"), line, after_word) or ""
        first_ref = _match_at(_REF, line, after_word + len(space))
        if first_ref is None:
            out.append(word)
            pos += len(word)
            continue

        # Emit the keyword and its first reference unchanged, then walk the rest
        # of the list, giving each later reference its own keyword.
        keyword = word.lower()
        out.append(word + space + first_ref)
        pos = after_word + len(space) + len(first_ref)

        while True:
            separator = _separator_at(line, pos)
            if separator is None:
                break
            ref = _match_at(_REF, line, pos + len(separator))
            if ref is None:
                # A separator not followed by a reference ends the list. This is
                # also what makes the rewrite idempotent: in
                # `closes #1, closes #2` the `, ` is followed by a keyword.
                break
            out.append(f"{separator}{keyword} {ref}")
            pos += len(separator) + len(ref)

    return "".join(out)


def normalize(text: str) -> str:
    """Normalize every line of `text`, preserving its line structure."""
    return "\n".join(normalize_line(line) for line in text.split("\n"))
